rules.js

```diff
@@ -0,0 +1,126 @@
+import { createReplaceRegex, defaultPlaceholders } from './regex_util.js';
+
+export function createRules({ match, replace, wds, webServerPort, compressed }) {
+  let ruleIdCounter = 1;
+
+  function createRegexReplace(searchPattern, replacePattern, useWDSBase) {
+    const replaceURL = new URL(replace);
+    const wdsBase = `${replaceURL.protocol}//${replaceURL.hostname}:${webServerPort}`;
+    const isWDS = wds === 'true';
+    const isCompressed = compressed === 'true';
+
+    const env = {
+      name: defaultPlaceholders.segment,
+      hash: defaultPlaceholders.hex,
+      ext: defaultPlaceholders.alphanumeric,
+      orig_url: match,
+      repl_url: isWDS && useWDSBase ? wdsBase : replace,
+      min: isCompressed ? '.min' : '',
+    };
+
+    const options = {
+      backReferenceSymbol: '\\',
+    };
+
+    return createReplaceRegex(searchPattern, replacePattern, env, options);
+  }
+
+  function createRedirectRule(searchPattern, replacePattern, useWDSBase, types) {
+    const { searchRegex, replaceRegex } = createRegexReplace(searchPattern, replacePattern, useWDSBase);
+    return {
+      id: ruleIdCounter++,
+      action: {
+        type: 'redirect',
+        redirect: {
+          regexSubstitution: replaceRegex,
+        },
+      },
+      condition: {
+        regexFilter: searchRegex,
+        resourceTypes: types,
+      },
+    };
+  }
+
+  function createCorsRule(pattern, types) {
+    const { searchRegex } = createRegexReplace(pattern, '', true);
+    return {
+      id: ruleIdCounter++,
+      action: {
+        type: 'modifyHeaders',
+        responseHeaders: [
+          {
+            header: 'Access-Control-Allow-Origin',
+            operation: 'set',
+            value: '*',
+          },
+          {
+            header: 'Access-Control-Allow-Headers',
+            operation: 'set',
+            value: '*',
+          },
+        ],
+      },
+      condition: {
+        regexFilter: searchRegex,
+        resourceTypes: types,
+      },
+    };
+  }
+
+  return [
+    // HMR updates
+    createRedirectRule(
+      '[orig_url]/static/[segment]/[name].[hash].hot-update.[ext]',
+      '[repl_url]/static/[segment]/[name].[hash].hot-update.[ext]',
+      true,
+      ['script', 'stylesheet', 'other', 'xmlhttprequest']
+    ),
+    // JS chunks
+    createRedirectRule(
+      '[orig_url]/static/[segment]/[name].[hash].js',
+      '[repl_url]/static/[segment]/[name][min].js',
```

Comment:
I'm a little bit skeptical of using a custom syntax parser tool, but I can't deny that this is really nice and readable